File: funcscript/block/LiteralBlock.cs
﻿using System.Text;
using funcscript.core;

namespace funcscript.block
{
    public class LiteralBlock : ExpressionBlock
    {
        public object Value;
        public LiteralBlock(object val)
        {
            Value = val;
        }

        public override string AsExpString(IFsDataProvider provider)
        {
            var sb = new StringBuilder();
            FuncScript.Format(sb, Value, null, true, false);
            return sb.ToString();
        }

        public override (object,CodeLocation) Evaluate(IFsDataProvider provider)
        {
            if (Value is ExpressionFunction exp)
            {
                lock (exp)
                {
                    exp.SetContext(provider);
                }
            }
            return (Value,this.CodeLocation);
        }
        public override IList<ExpressionBlock> GetChilds()
        {
            return new ExpressionBlock[0];
        }
        public override string ToString()
        {
            if (Value == null)
                return "";
            return Value.ToString();
        }

    }

}

File: js-port/funscscript-js/src/block/literal-block.js
const { ExpressionBlock } = require('./expression-block');
const { ensureTyped } = require('../core/value');
const { ExpressionFunction } = require('../core/expression-function');

class LiteralBlock extends ExpressionBlock {
  constructor(value, position = 0, length = 0) {
    super(position, length);
    this.value = ensureTyped(value);
  }

  evaluate(provider) {
    const raw = this.value[1];
    if (raw instanceof ExpressionFunction) {
      raw.setContext(provider);
    }
    return this.value;
  }

  asExpressionString() {
    const val = this.value[1];
    if (val === null) {
      return 'null';
    }
    return String(val);
  }
}

module.exports = {
  LiteralBlock
};
